state/txindex/kv/kv_test.go

```diff
@@ -11,30 +12,103 @@ import (
 	"github.com/tendermint/tendermint/state/txindex"
 	"github.com/tendermint/tendermint/types"
 	db "github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/pubsub/query"
 )
 
 func TestTxIndex(t *testing.T) {
 	indexer := &TxIndex{store: db.NewMemDB()}
 
 	tx := types.Tx("HELLO WORLD")
-	txResult := &types.TxResult{1, 0, tx, abci.ResponseDeliverTx{Data: []byte{0}, Code: abci.CodeType_OK, Log: ""}}
+	txResult := &types.TxResult{1, 0, tx, abci.ResponseDeliverTx{Data: []byte{0}, Code: abci.CodeType_OK, Log: "", Tags: []*abci.KVPair{}}}
 	hash := tx.Hash()
 
 	batch := txindex.NewBatch(1)
-	if err := batch.Add(*txResult); err != nil {
+	if err := batch.Add(txResult); err != nil {
 		t.Error(err)
 	}
-	err := indexer.AddBatch(batch)
-	require.Nil(t, err)
+	err := indexer.AddBatch(batch, []string{})
+	require.NoError(t, err)
 
 	loadedTxResult, err := indexer.Get(hash)
-	require.Nil(t, err)
+	require.NoError(t, err)
 	assert.Equal(t, txResult, loadedTxResult)
+
+	tx2 := types.Tx("BYE BYE WORLD")
+	txResult2 := &types.TxResult{1, 0, tx2, abci.ResponseDeliverTx{Data: []byte{0}, Code: abci.CodeType_OK, Log: "", Tags: []*abci.KVPair{}}}
+	hash2 := tx2.Hash()
+
+	err = indexer.Index(txResult2, []string{})
+	require.NoError(t, err)
+
+	loadedTxResult2, err := indexer.Get(hash2)
+	require.NoError(t, err)
+	assert.Equal(t, txResult2, loadedTxResult2)
+}
+
+func TestTxSearch(t *testing.T) {
+	indexer := &TxIndex{store: db.NewMemDB()}
+
+	tx := types.Tx("HELLO WORLD")
+	tags := []*abci.KVPair{
+		&abci.KVPair{Key: "account.number", ValueType: abci.KVPair_INT, ValueInt: 1},
+		&abci.KVPair{Key: "account.owner", ValueType: abci.KVPair_STRING, ValueString: "Ivan"},
+		&abci.KVPair{Key: "not_allowed", ValueType: abci.KVPair_STRING, ValueString: "Vlad"},
+	}
+	txResult := &types.TxResult{1, 0, tx, abci.ResponseDeliverTx{Data: []byte{0}, Code: abci.CodeType_OK, Log: "", Tags: tags}}
+	hash := tx.Hash()
+
+	allowedTags := []string{"account.number", "account.owner", "account.date"}
+	err := indexer.Index(txResult, allowedTags)
+	require.NoError(t, err)
+
+	testCases := []struct {
+		q             string
+		expectError   bool
+		resultsLength int
+		results       []*types.TxResult
```

Comment:
lets drop `results` and `expectError` here for readability. resultsLength in this test is always {0, 1}, so we can just refer to `txResult` in the test loop, and there's always no error.